TCP2 TX data path refactoring

jukkar opened the pull request:
Refactor TCP2 TX data path. After this it should be more easy to implement retransmit logic.
@ozhuraki the commits might miss some cases with test protocol logic.

Edit: I left out commits from #24852 but that one is required and must be applied before this PR.

jukkar committed:
net: tcp2: Refactoring data sending and receive path

Change how TCP data is being sent and received. In sending side,
the data is put to one list net_buf's, and in receiving side we
pull the data after receiving ACKs for them.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp2: Randomize initial sequence number

The initial seq number for a connection should be randomized
to prevent easy guesses.

If network test framework is enabled, then do not randomize the
initial seq so that the tests do not fail.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/ip/tcp2.c:
```diff
@@ -834,6 +940,45 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
+		u32_t acked_len = th_ack(th) - conn->last_seq;
+		int ret;
+
+		if (conn->pending_data && acked_len) {
```
maybe test the contrary and return if it matches. Then the code block below will be indented less (there is 4 nested if now)

jukkar replied:
Sure, that makes sense.

tbursztyka commented on subsys/net/ip/tcp2.c:
```diff
@@ -480,89 +631,43 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
+		 * - the send function will figure out how much data to
+		 *   send according to send window size and mss
+		 */
 		va_list ap;
```
Could you take the opportunity to remove the usage of variable argument list on this function? It would take a net_pkt pointer as argument after flags, and that's it. It's way more readable that way.

jukkar replied:
That is the ultimate intention, the va_list is not really needed here. But I would like to avoid changing it in this PR in order to minimize the changes.

tbursztyka replied:
ok

ozhuraki commented on subsys/net/ip/tcp2_priv.h:
```diff
@@ -180,3 +184,14 @@ struct tcp { /* TCP connection */
+ *
+ * @return Return a random TCP sequence number
+ */
+static inline u32_t tcp_init_isn(void)
```
Shall we just reuse sys_rand32_get() in place? Currently the TTCN-3 sanity check needs it not randomized, so under #ifdef CONFIG_NET_TEST_PROTOCOL

jukkar replied:
Sure, I will add relevant code.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -924,15 +1073,19 @@ static void tcp_in(struct tcp *conn, struct net_pkt *pkt)
 				conn_ack(conn, + len);
 				tcp_out(conn, ACK);
-			} else if (th_seq(th) < conn->ack) {
+			} else if (net_tcp_seq_greater(conn->ack,
```
Shall we just use simple '>' here?

jukkar replied:
Wouldn't it fail if the numbers have wrapped?

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -832,6 +940,45 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
+		return;
+	}
+
+	if (!net_tcp_seq_greater(th_ack(th), conn->seq)) {
```
Shall we use a simpler check 'conn->seq >= th_ack(th)' here?

jukkar replied:
I am not sure how we could do that, the question is how would we catch the u32_t wraps then.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -267,6 +274,145 @@ int net_tcp_unref(struct net_context *context)
+		conn->send_win, conn->pending_len,
+		conn->seq, conn->ack, net_pkt_get_len(conn->pending_data));
+
+	pkt = net_pkt_alloc_with_buffer(conn->iface, len,
```
Let's use tcp_pkt_alloc() here, so we can track it when the test protocol is enabled.

jukkar replied:
Sure, that can be done.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -480,89 +631,43 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
 
-		len = net_pkt_get_len(data_pkt);
+		if (!conn->pending_data) {
+			conn->pending_data = net_pkt_alloc(TCP_PKT_ALLOC);
```
Let's use tcp_pkt_alloc() here, so we can track it when the test protocol is enabled.

jukkar replied:
Hmm, here that cannot be done as we do not allocate any buffers for the net_pkt.

ozhuraki replied:
@jukkar 
> Hmm, here that cannot be done as we do not allocate any buffers for the net_pkt.

We can update tcp_pkt_alloc() to handle this case.

jukkar replied:
That is probably possible, I will investigate.

jukkar replied:
I changed the code to support this.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -832,6 +940,45 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
+		return;
+	}
+
+	if (!net_tcp_seq_greater(th_ack(th), conn->seq)) {
```
Shall we even move this check into tcp_in()?

jukkar replied:
Sure, why not

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -208,6 +210,11 @@ static void tcp_send_queue_flush(struct tcp *conn)
 	}
+
+	if (conn->pending_data) {
+		net_pkt_unref(conn->pending_data);
```
Let's use tcp_pkt_unref()?

jukkar replied:
As we did not allocate the pending_data using tcp_pkt_alloc() we cannot use tcp_pkt_unref() here. Not sure if this is a big issue as the conn->pending_data is there for the whole duration of the connection. So it is not allocated/freed constantly.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -832,6 +940,45 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
+			net_pkt_set_overwrite(conn->pending_data, true);
+
+			ret = net_pkt_pull(conn->pending_data, acked_len);
+			if (ret < 0 && conn->pending_len != (int)acked_len) {
```
Would it be a good idea to check error cases with pending_len, acked_len before net_pkt_pull()?

jukkar replied:
Only thing to check is to calculate the length of the pending_data. IMHO this is too costly to do, as the pending_data can be quite large and using net_pkt_get_len() for each incoming ACK is heavy. If we would keep track of pending_data length separately, then it would be much better. I suggest we leave this to future patches.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -267,6 +274,145 @@ int net_tcp_unref(struct net_context *context)
+	return -EINVAL;
+}
+
+static void tcp_send_queued_data(struct k_work *work)
```
Would it be a good idea to rework/change tcp_send_queued_data() into tcp_add_pending/queued_data(), so it just adds the pending data into packet and to reuse tcp_out() practically intact?

jukkar replied:
Yes, my idea was also change this so that if we are sending ACK, we could check if there is any data that we could send in the same go. But I did not want to start to change too much here, it was quite a task to get this far already.

jukkar commented:
Updated according to comments.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -832,6 +936,40 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
+		}
+	}
+
+	conn->last_seq += acked_len;
```
I am not sure, I follow regarding the purpose of last_seq. Could you please try your changes with the TTCN-3 sanity check?

jukkar replied:
> I am not sure, I follow regarding the purpose of last_seq

I had issue when trying to use conn->seq, it was updated too early to be useful. But after these latest refactorings, perhaps it works now. I will try that.

jukkar replied:
If I remove the conn->last_seq, and use conn->seq directly, then we start to get these kind of issues (if we sent long packet and receiving only ACK to small portition of waiting data):
```
[00:00:29.700,000] <dbg> net_echo_server_sample.handle_data: TCP (IPv6): Received and replied with 640 bytes
[00:00:29.700,000] <dbg> net_tcp.tcp_in: (0x08081090): ACK,PSH Seq=2560866935 Ack=334960246 Len=592 [ESTABLISHED Seq=334960886 Ack=256
[00:00:29.700,000] <dbg> net_tcp.update_acked_data: (0x08081090): Peer acked 4294966657 bytes
[00:00:29.700,000] <err> net_tcp: Trying to ack more data (-639) that was in the queue (671)
```
So the idea of last_seq was to keep track of what is the position in the data stream that we are waiting ack. So in theory we should be able to use conn->seq for that but there is now some complex interaction in the code but I am a bit lost where to look.

ozhuraki replied:
> So in theory we should be able to use conn->seq for that but there is now some complex interaction in the code but I am a bit lost where to look.

One way would be not to update conn->seq as it’s done in this patch, but to use pending_len to calculate the sequence number for the subsequent data packets.

jukkar replied:
Could be, although everything seems to work now and I would like to avoid this optimization at this point.

tbursztyka commented on subsys/net/ip/tcp2.c:
```diff
@@ -267,6 +272,143 @@ int net_tcp_unref(struct net_context *context)
+		return -ENOBUFS;
+	}
+
+	memset(th, 0, sizeof(struct tcphdr));
```
the function is already setting more than half of the attributes, let's do like in ipv4/ipv6: set all attributes specifically and in the order they appear (nevermind for for th_x2/th_off)

jukkar replied:
Yep, something like that can be done. But I would like to avoid changing too much here with this commit in order not to cause unrelated changes, this function was just moved around in order to avoid forward declaration.

tbursztyka replied:
ok

jukkar committed:
tests: net: tcp2: Avoid trying to free net_pkt twice

For success cases, the net_pkt is handled by the IP stack
so no need to free the net_pkt. The double free is not causing
any issues as the net_pkt code will just ignore the request but
we should avoid doing it because of error prints that are printed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Lets close this and do all the work in #25029 instead.
